blueman/main/DbusService.py

```diff
@@ -1,56 +1,145 @@
+            self.register()
+
+    def _handle_method_call(self, _connection, sender, _path, interface_name, method_name, parameters, invocation):
+        try:
+            try:
+                _arguments, result_signature, method, options = self._methods[method_name]
+            except KeyError:
+                logging.warning('Unhandled method: %s' % method_name)
+                invocation.return_error_literal(Gio.dbus_error_quark(), Gio.DBusError.UNKNOWN_METHOD,
+                                                "No such method on interface: %s.%s" % (interface_name, method_name))
+
+            def ok(*result):
+                invocation.return_value(self._prepare_arguments(result_signature, result))
+
+            args = parameters.unpack()
+            if "sender" in options:
+                args += (sender,)
+            if "async" in options:
+                method(*(args + (ok, lambda exception: self._return_dbus_error(invocation, exception))))
+            else:
+                ok(method(*args))
+        except Exception as e:
+            self._return_dbus_error(invocation, e)
 
-
-class MethodAlreadyExists(Exception):
-    pass
-
-
-class DbusService(dbus.service.Object):
-    def __init__(self, name, path, bus=dbus.SessionBus):
-        self.bus = bus()
-        self.bus.request_name(name)
-
-        super(DbusService, self).__init__(self.bus, path)
-
-    def add_definitions(self, instance):
-        setattr(instance, 'locations', list(self.locations))
-
-        for name, func in self._definitions(instance):
-            if name in self.__class__.__dict__:
-                raise MethodAlreadyExists
-
-            self._add_wrapper(getattr(instance, name))
-
-        self.__class__._refresh_dbus_registration()
-
-    def _add_wrapper(self, method):
-        def wrapper(*args, **kwargs):
-            return method(*args[1:], **kwargs)
-
-        wrapper.__name__ = method.__name__
-
-        for key, value in method.__dict__.items():
-            if key.startswith('_dbus_'):
-                setattr(wrapper, key, value)
-
-        setattr(self.__class__, method.__name__, wrapper)
-
-    def remove_definitions(self, obj):
-        for name, func in self._definitions(obj):
-            delattr(self.__class__, name)
-
-        self.__class__._refresh_dbus_registration()
-
-    @classmethod
-    def _refresh_dbus_registration(cls):
-        cls.__class__.__init__(cls, cls.__name__, cls.__bases__, cls.__dict__)
+    @staticmethod
+    def _return_dbus_error(invocation, data):
+        if isinstance(data, DbusError):
+            invocation.return_dbus_error(data.name, data.message)
+        else:
+            if isinstance(data, Exception):
+                et, ev, etb = sys.exc_info()
+                if ev is data:
+                    message = "".join(traceback.format_exception(et, ev, etb))
+                else:
+                    message = "".join(traceback.format_exception_only(data.__class__, data))
```

Comment:
This is taken from dbus-python (https://gitlab.freedesktop.org/dbus/dbus-python/blob/master/dbus/service.py#L285).

The message e.g. for our `DBusFailedError` thus matches the old one now (`blueman.bluez.errors.DBusFailedError:  Host is down`).